File: examples/canonical-auth/go/tests/token_test.go
package tests

import (
	"testing"
	"time"

	"canonical-auth/internal/auth"
	"canonical-auth/internal/domain"

	"github.com/golang-jwt/jwt/v5"
)

func setupTestStore(t *testing.T) *domain.InMemoryStore {
	store, err := domain.LoadFromFixtures("../../domain/fixtures")
	if err != nil {
		t.Fatalf("failed to load fixtures: %v", err)
	}
	return store
}

func TestIssueToken_ClaimsShape(t *testing.T) {
	t.Setenv("TENANTCORE_TEST_SECRET", "test-secret")
	store := setupTestStore(t)
	// alice@acme.example is a tenant_member and is in a group
	user, ok := store.GetUserByEmail("alice@acme.example")
	if !ok {
		t.Fatal("user not found")
	}

	key := []byte("test-secret")
	tokenString, err := auth.IssueToken(user, store, key)
	if err != nil {
		t.Fatalf("failed to issue token: %v", err)
	}

	token, err := jwt.ParseWithClaims(tokenString, &auth.Claims{}, func(token *jwt.Token) (interface{}, error) {
		return key, nil
	}, jwt.WithValidMethods([]string{"HS256"}), jwt.WithIssuer("tenantcore-auth"), jwt.WithAudience("tenantcore-api"))

	if err != nil {
		t.Fatalf("failed to parse token: %v", err)
	}

	claims, ok := token.Claims.(*auth.Claims)
	if !ok || !token.Valid {
		t.Fatal("invalid token claims")
	}

	if claims.Subject != user.ID {
		t.Errorf("expected sub %s, got %s", user.ID, claims.Subject)
	}
	if claims.TenantID == nil {
		t.Error("expected tenant_id not to be nil")
	}
	if claims.TenantRole != "tenant_member" {
		t.Errorf("expected tenant_role tenant_member, got %s", claims.TenantRole)
	}
	if len(claims.Groups) == 0 {
		t.Error("expected groups to be populated")
	}
	if claims.ACLVer != user.ACLVer {
		t.Errorf("expected acl_ver %d, got %d", user.ACLVer, claims.ACLVer)
	}
}

func TestIssueToken_Expiry(t *testing.T) {
	t.Setenv("TENANTCORE_TEST_SECRET", "test-secret")
	store := setupTestStore(t)
	user, _ := store.GetUserByEmail("alice@acme.example")
	key := []byte("test-secret")
	tokenString, err := auth.IssueToken(user, store, key)
	if err != nil {
		t.Fatalf("failed to issue token: %v", err)
	}

	token, _ := jwt.ParseWithClaims(tokenString, &auth.Claims{}, func(token *jwt.Token) (interface{}, error) {
		return key, nil
	}, jwt.WithValidMethods([]string{"HS256"}))

	claims := token.Claims.(*auth.Claims)
	exp := claims.ExpiresAt.Time
	iat := claims.IssuedAt.Time
	diff := exp.Sub(iat)

	if diff != 15*time.Minute {
		t.Errorf("expected 15 min expiry, got %v", diff)
	}
}

func TestIssueToken_Permissions(t *testing.T) {
	t.Setenv("TENANTCORE_TEST_SECRET", "test-secret")
	store := setupTestStore(t)
	user, _ := store.GetUserByEmail("alice@acme.example")
	key := []byte("test-secret")
	tokenString, err := auth.IssueToken(user, store, key)
	if err != nil {
		t.Fatalf("failed to issue token: %v", err)
	}

	token, _ := jwt.ParseWithClaims(tokenString, &auth.Claims{}, func(token *jwt.Token) (interface{}, error) {
		return key, nil
	}, jwt.WithValidMethods([]string{"HS256"}))

	claims := token.Claims.(*auth.Claims)

	effectivePerms := store.GetEffectivePermissions(user.ID)

	if len(claims.Permissions) != len(effectivePerms) {
		t.Errorf("expected %d permissions, got %d", len(effectivePerms), len(claims.Permissions))
	}
}

File: examples/canonical-auth/typescript/tests/token.test.ts
import assert from "node:assert/strict";
import { describe, test } from "node:test";
import { issueToken, getVerificationKey } from "../src/auth/token.ts";
import { store } from "../src/domain/store.ts";
import { jwtVerify } from "jose";

process.env.TENANTCORE_TEST_SECRET = "test-secret-12345";
process.env.TENANTCORE_TEST_ALG = "HS256";

describe("Token Unit Tests", () => {
  test("issueToken produces a verifiable JWT", async () => {
    const user = store.getUserByEmail("alice@acme.example");
    if (!user) throw new Error("Fixture user missing");

    const token = await issueToken(user, store);
    assert.ok(token);

    const key = await getVerificationKey();
    const { payload } = await jwtVerify(token, key, {
      issuer: "tenantcore-auth",
      audience: "tenantcore-api",
    });

    assert.equal(payload.iss, "tenantcore-auth");
    assert.equal(payload.aud, "tenantcore-api");
    assert.equal(payload.sub, user.id);
    assert.equal((payload.exp as number) - (payload.iat as number), 900);
    assert.equal(payload.tenant_id, user.tenant_id);
    assert.equal(payload.tenant_role, "tenant_member");
    assert.equal(Array.isArray(payload.groups), true);
    assert.equal(Array.isArray(payload.permissions), true);
    assert.equal(payload.acl_ver, user.acl_ver);
  });

  test("super_admin token shape", async () => {
    const user = store.getUserByEmail("superadmin@tenantcore.dev");
    if (!user) throw new Error("Fixture user missing");

    const token = await issueToken(user, store);
    
    const key = await getVerificationKey();
    const { payload } = await jwtVerify(token, key, {
      issuer: "tenantcore-auth",
      audience: "tenantcore-api",
    });

    assert.equal(payload.tenant_id, undefined);
    assert.equal(payload.tenant_role, "super_admin");
  });
});
